src/coreclr/vm/object.cpp

```diff
@@ -1464,95 +1464,86 @@ void StackTraceArray::CheckState() const
 
-    if (!m_array)
-    {
-        SetArray(I1ARRAYREF(AllocatePrimitiveArray(ELEMENT_TYPE_I1, static_cast<DWORD>(raw_size))));
-        SetSize(0);
-        SetObjectThread();
-    }
-    else
+    if (raw_size.IsOverflow() || !FitsIn<DWORD>(raw_size.Value()))
     {
-        if (Capacity() >= raw_size)
-            return;
-
-        // allocate a new array, copy the data
-        size_t new_capacity = Max(Capacity() * 2, raw_size);
-
-        _ASSERTE(new_capacity >= grow_size * sizeof(StackTraceElement) + sizeof(ArrayHeader));
-
-        I1ARRAYREF newarr = (I1ARRAYREF) AllocatePrimitiveArray(ELEMENT_TYPE_I1, static_cast<DWORD>(new_capacity));
-        memcpyNoGCRefs(newarr->GetDirectPointerToNonObjectElements(),
-                       GetRaw(),
-                       Size() * sizeof(StackTraceElement) + sizeof(ArrayHeader));
-
-        SetArray(newarr);
+        EX_THROW(EEMessageException, (kOverflowException, IDS_EE_ARRAY_DIMENSIONS_EXCEEDED));
     }
+   
+    SetArray(I1ARRAYREF(AllocatePrimitiveArray(ELEMENT_TYPE_I1, static_cast<DWORD>(raw_size.Value()))));
+    SetSize(0);
+    SetKeepAliveItemsCount(0);
+    SetObjectThread();
 }
 
-void StackTraceArray::EnsureThreadAffinity()
+size_t StackTraceArray::Capacity() const
 {
     WRAPPER_NO_CONTRACT;
-
     if (!m_array)
-        return;
+    {
+        return 0;
+    }
+
+    return (m_array->GetNumComponents() - sizeof(ArrayHeader)) / sizeof(StackTraceElement);
+}
 
-    if (GetObjectThread() != GetThreadNULLOk())
+// Compute the number of methods in the stack trace that can be collected. We need to store keepAlive
+// objects (Resolver / LoaderAllocator) for these methods.
+uint32_t StackTraceArray::ComputeKeepAliveItemsCount()
+{
+    LIMITED_METHOD_CONTRACT;
+
+    uint32_t count = 0;
+    for (uint32_t i = 0; i < Size(); i++)
     {
-        // object is being changed by a thread different from the one which created it
-        // make a copy of the array to prevent a race condition when two different threads try to change it
-        StackTraceArray copy;
-        GCPROTECT_BEGIN(copy);
-            copy.CopyFrom(*this);
-            this->Swap(copy);
-        GCPROTECT_END();
+        if ((*this)[i].flags & STEF_KEEPALIVE)
+        {
+            count++;
+        }
     }
+
+    return count;
 }
 
-// Deep copies the stack trace array
-void StackTraceArray::CopyFrom(StackTraceArray const & src)
+uint32_t StackTraceArray::CopyDataFrom(StackTraceArray const & src)
 {
     CONTRACTL
     {
-        THROWS;
-        GC_TRIGGERS;
+        NOTHROW;
+        GC_NOTRIGGER;
         MODE_COOPERATIVE;
-        INJECT_FAULT(ThrowOutOfMemory(););
         PRECONDITION(IsProtectedByGCFrame((OBJECTREF*)this));
         PRECONDITION(IsProtectedByGCFrame((OBJECTREF*)&src));
     }
     CONTRACTL_END;
 
-    m_array = (I1ARRAYREF) AllocatePrimitiveArray(ELEMENT_TYPE_I1, static_cast<DWORD>(src.Capacity()));
-
-    Volatile<size_t> size = src.Size();
+    uint32_t size = src.Size();
```

Comment:
We should use volatile read for the Size. Otherwise, the C++ compiler is free to duplicate the read and use one `size` for value for the memcpy and return a different value from the method.

Reply:
I have thought you were suggesting before that I get rid of the Volatile here, so I am bit confused.

Reply:
I would:
- Get rid of the volatile local variable. It does not make sense to tell the compiler to issue read/write barriers when reading/writing stack. It is deoptimization.
- Change the read in Size() method to be volatile read, so that it is not reordered. This volatile read is counterpart of volatile write in `StackTraceArray::Append` (it is not actual volatile write, but MemoryBarrier + regular write that is equivalent in this situation). In general, volatile reads and volatile writes in lock-free algorithm have to come in pairs on producing/consuming sides.

Reply:
Ok, I have added a commit with that change. I've modified the GetSize / SetSize and added such treatment to the keep alive count accessors too. And removed the MemoryBarrier from the StackTraceArray::Append where it is no longer needed after this change.